docs/downloader.rst

```diff
@@ -0,0 +1,169 @@
+-----------------
+
+The ``DownloaderService`` is responsible for deciding which Snapshots to download.
+The capability strings of Snapshots to download are given to ``RemoteSnapshotCacheService`` which downloads a Snapshot and all its relevant parents into memory.
+Once a Snapshot is downloaded it is passed to the ``MagicFolderUpdaterService`` which stages the actual content and determins if the change is an "overwrite" or a "conflict" and changes the local filesystem accordingly.
+
+Downloading new Snapshots from other participants ultimately causes changes to the local filesystem (in the magic-folder).
+
+
+What Snapshots to Download
+--------------------------
+
+Each configured magic-folder has a ``collective_dircap`` which is a Tahoe-LAFS Directory Capability ("dircap") for the list of participants.
+If this dircap is writable then this device is the administrator (and the only device which can modify the participants).
+
+In either case, the capability can be downloaded.
+It will be a Tahoe-LAFS directory containing a series of sub-directories; these are the participants.
+The directory name is the Participant name and points at a directory read-capability where all the files in their magic-folder are stored.
+This directory is known as the "Personal DMD" (where DMD stands for "Distributed Mutable Directory").
+
+The entries in a user's Personal DMD are flat (no subdirectories) and point a (mangled) relative path-name to a Snapshot.
+
+So, Snapshots to download are discovered by:
+
+- reading the Collective DMD
+- for each user in it:
+  - read their Personal DMD
+  - for each entry in the Personal DMD:
+    - queue the Snapshot for caching, including all parents
+    - after download, queue the Snapshot for updates
+
+Note that we don't always download *every* parent; we can stop early if a common ancestor is found.
+
+
+Downloading Snapshots
+---------------------
+
+The ``RemoteSnapshotCacheService`` awaits capability-strings of Snapshots to download.
+For each one, the function ``create_snapshot_from_capability`` is used to download the capability and return a ``RemoteSnapshot`` instance.
+This ``RemoteSnapshot`` is stored in memory.
+
+Additionally, all parents of the Snapshot are also cached.
+We can stop caching parents once we find a "common ancestor"; this means a parent in the remote snapshot that matches the one in our Personal DMD.
+
+Conflict Resolution is described in :ref:`Multi-party Conflict Detection` under the Leif's Design section.
+Briefly: a ``RemoteSnapshot`` is traced through its parents until a common ancestor is found.
+If the new Snapshot is a descendant of our latest Snapshot for that name, it's an overwrite.
+If it is not, there is a conflict (unless we don't yet have that name at all, then it's a creation).
+
+
+On Overwrite
+------------
+
+The ``content`` of the ``RemoteSnapshot`` is downloaded and moved into place in our Magic Folder.
+Our Personal DMD is updated to point at this Snapshot.
```

Comment:
I think we need to be explicit about the order things are done here, and how we ensure that we don't have inconsistencies or cause spurious conflicts if the process is racing creating a local snapshot.

In particular, we don't want to create a snapshot of a partially written remote file. This isn't *currently* a problem, but as I mention below, we may want to change it to avoid blocking while writing a large file, in which case it will definitely be an issue.

Reply:
The order is as-written there. Does it need more words?

There is no Snapshot created here, just the pointer in the Personal DMD is updated (but, yes, it shouldn't be updated until the on-disk data matches what is in that Snapshot).

Reply:
> The order is as-written there. Does it need more words?

We have pseudocode elsewhere for things that have a specific order. It wasn't obvious to me that this description in paragraph from was meant to be ordered/.

A couple of things that would be useful to have:
- where in the process we update the db, and with what
- some explanation of why this particular order. I assume it has to do with:
  - failing at any point in this process is recoverable (why, what invariants are being maintained that we can recover)
  - how it interacts with other process that may be going on at the same time, like a snapshot being taken at the same time, and the order that process takes as compared with this one mean that both maintain a consistent state.

>> In particular, we don't want to create a snapshot of a partially written remote file. This isn't currently a problem, but as I mention below, we may want to change it to avoid blocking while writing a large file, in which case it will definitely be an issue.
>
> There is no Snapshot created here, just the pointer in the Personal DMD is updated (but, yes, it shouldn't be updated until the on-disk data matches what is in that Snapshot).

What I'm talking about is if `LocalSnapshotCreator.store_local_snapshot` is running at the same time as `MagicFolderUpdaterService._process`. What I am asking about is if we are going to get a sensible result if the steps of those two functions are interleaved arbitrarily, are we going to get a reasonable result? One particular worry I have is whether we are going to get a spurious conflict because  `LocalSnapshotCreator.store_local_snapshot` creates a local snapshot of the new file data, after `MagicFolderUpdaterService._process` has written the new file, but before we've recorded any information about it.

Reply:
Okay, I added words about when the database is updated.

I believe your worry won't happen: we move the file into place and update our database before doing any more async work (which is the "update our Personal DMD" step). A new LocalSnapshot can only be created before we move the file or after we've updated the database. If it happens before, we'll have the upstream parent (which is fine). If it happens right after the database update but before the Personal DMD update, we'll have our new snapshot as parent.

Reply:
> I believe your worry won't happen: [...]

We could crash between those two points, which would have similar issues. For the non-failure case, if it is important that those two steps are done synchronously (in other words, protected by the implicit global lock of running in the reactor thread), then that should be called out.

Reply:
okay, I added a comment.